interface/main/main_screen.php

```diff
@@ -90,87 +90,136 @@ function generate_html_end()
 $errormsg = '';
 
 ///////////////////////////////////////////////////////////////////////
-// Begin code to support U2F.
+// Begin code to support U2F and APP Based TOTP logic.
 ///////////////////////////////////////////////////////////////////////
 
 $regs = array();          // for mapping device handles to their names
 $registrations = array(); // the array of stored registration objects
 $res1 = sqlStatement(
-    "SELECT a.name, a.var1 FROM login_mfa_registrations AS a " .
-    "WHERE a.user_id = ? AND a.method = 'U2F' ORDER BY a.name",
+    "SELECT a.name, a.method, a.var1 FROM login_mfa_registrations AS a " .
+    "WHERE a.user_id = ? ORDER BY a.name",
     array($_SESSION['authId'])
 );
 while ($row1 = sqlFetchArray($res1)) {
     $regobj = json_decode($row1['var1']);
     $regs[json_encode($regobj->keyHandle)] = $row1['name'];
     $registrations[] = $regobj;
+    $registrations["method"] = $row1['method'];
+    if ($row1['method'] == 'TOTP') {
+        $registrations['secret'] = $row1['var1'];
+    }
```

Comment:
Note you are repeating the above 3 lines below (keeping it below seems more clear)

Reply:
This part doesn't look right. You should allow for both U2F and TOTP registrations, however you have only one place to store a method. Also $registrations is specific to U2F and you are putting things in it that are not related to U2F. I think what you need is a separate block of code for TOTP.

Reply:
Brady Miller: I chose to add this here because the U2F code is calling upon the reg object and storing it into the array afterwards. I did not want to mess with the alterations to the array beforehand since I don't have a U2F device to test with and am not sure what is being stored there.

Sunset Systems: I chose to not separate the blocks this way because I felt that it added an extra SQL call to get redundant information. I can store this information into another array, but as I mentioned before I'm not sure what is being stored in the registrations array beforehand and can't test to see if I accidentally break that code. You are correct in saying that only one U2F or TOTP registration will be called.

If you'd like, I can alter the code in one of two ways.

1) I can default the MFA method originally used to use U2F with an option to switch to the TOTP method in the view. I think this would make sense since U2F is probably more secure. This would only require another GET/POST parameter to determine which method is being used. This would also not require the alteration of the database table login_mfa_registrations. I could also give them the option to switch back as well.

2) I can create an additional column in the login_mfa_registrations table with a boolean value of primary and then pull the primary method from there. This would require some UI modification to the MFA registration view as well to allow a user to select which method they would like to use as their primary MFA method. Of course, there would also be a toggle to switch back and forth between U2F and TOTP methods.

Or I can change it in a third way of your suggestion.

Reply:
One thing you might do is modify the current page with the Authenticate button, changing it to "Authenticate U2F", and then also add your TOTP input field and an "Authenticate TOTP" button. Add suitable explanatory text as appropriate. In this way there's no need to indicate which is primary because the user just picks what they want with no extra clicks. And of course if there's only one registered method then you just show the controls for that. How does that sound?

Reply:
Regarding the $registrations variable in the code, suggest leaving it alone and create new variables for TOTP data. You can definitely use the existing login_mfa_registrations table.

Reply:
Ah, yes, I like this method better! If you're okay with it, I'll proceed.

Reply:
When close to done, I'm sure Brady can create a demo site and then I can test with both methods.